packages/flutter/lib/src/widgets/widget_inspector.dart

```diff
@@ -146,10 +149,203 @@ class WidgetInspectorService {
+    @required Object callback(List<String> args),
+  }) {
+    registerServiceExtension(
+      name: name,
+      callback: (Map<String, String> parameters) {
+        const String argPrefix = 'arg';
+        final List<String> args = <String>[];
+        parameters.forEach((String name, String value) {
+          if (name.startsWith(argPrefix)) {
+            final int index = int.parse(name.substring(argPrefix.length));
+            if (index >= args.length) {
+              args.length = index + 1;
+            }
+            args[index] = value;
+          }
+        });
+        return callback(args);
+      },
+    );
+  }
+
+  /// Called to register service extensions.
+  ///
+  /// Service extensions are only exposed when the observatory is
+  /// included in the build, which should only happen in checked mode
+  /// and in profile mode.
+  ///
+  /// See also:
+  ///
+  ///  * <https://github.com/dart-lang/sdk/blob/master/runtime/vm/service/service.md#rpcs-requests-and-responses>
+  void initServiceExtensions() {
+    assert(!_debugServiceExtensionsRegistered);
+    _debugServiceExtensionsRegistered = true;
+
+    _registerSignalServiceExtension(
+      name: 'disposeAllGroups',
+      callback: disposeAllGroups,
+    );
+    _registerObjectGroupServiceExtension(
+      name: 'disposeGroup',
+      callback: disposeGroup,
+    );
+    _registerSignalServiceExtension(
+      name: 'isWidgetTreeReady',
+      callback: isWidgetTreeReady
+    );
+    _registerServiceExtensionWithArg(
+      name: 'disposeId',
+      callback: disposeId,
+    );
+    _registerServiceExtensionVarArgs(
+      name: 'setPubRootDirectories',
+      callback: setPubRootDirectories,
+    );
+    _registerServiceExtensionWithArg(
+      name: 'setSelectionById',
+      callback: setSelectionById,
+    );
+    _registerServiceExtensionWithArg(
+      name: 'getParentChain',
+      callback: _getParentChain,
+    );
+    _registerServiceExtensionWithArg(
+      name: 'getProperties',
+      callback: _getProperties,
+    );
+    _registerServiceExtensionWithArg(
+      name: 'getChildren',
+      callback: _getChildren,
+    );
+    _registerObjectGroupServiceExtension(
+      name: 'getRootWidget',
+      callback: _getRootWidget,
+    );
+    _registerObjectGroupServiceExtension(
+      name: 'getRootRenderObject',
+      callback: _getRootRenderObject,
+    );
+    _registerServiceExtensionWithArg(
+      name: 'getSelectedRenderObject',
+      callback: _getSelectedRenderObject,
+    );
+    _registerServiceExtensionWithArg(
+      name: 'getSelectedWidget',
+      callback: _getSelectedWidget,
+    );
+    _registerSignalServiceExtension(
+      name: 'isWidgetCreationTracked',
+      callback: isWidgetCreationTracked,
+    );
+  }
+
   /// Clear all InspectorService object references.
   ///
   /// Use this method only for testing to ensure that object references from one
   /// test case do not impact other test cases.
+  @visibleForTesting
```

Comment:
I would make all these @protected, then just have a subclass in the tests, and have the subclass be the one you construct. You could even go further and require that the only way to extend this class is as a mixin, then you could hide the constructor on this version.

Reply:
looks like you're actually a lot of the way there already actually.

Reply:
I've followed both those suggestions. The only way to extend is now a mixin and the tests are executed in a context where they can access protected members.